crates/concrete_check/src/linearity_check.rs

```diff
@@ -0,0 +1,1086 @@
+                },
+            );
+        }
+    }
+    // Example of updating the state table
+    fn update_info(&mut self, var: &str, info: VarInfo) {
+        self.vars.insert(var.to_string(), info);
+    }
+
+    // Remove a variable from the state table
+    fn _remove_entry(&mut self, var: &str) {
+        self.vars.remove(var);
+    }
+
+    fn get_info_mut(&mut self, var: &str) -> Option<&mut VarInfo> {
+        if !self.vars.contains_key(var) {
+            self.vars.insert(
+                var.to_string(),
+                VarInfo {
+                    ty: "".to_string(),
+                    depth: 0,
+                    state: VarState::Unconsumed,
+                },
+            );
+            tracing::debug!(
+                "Variable {} not found in state table. Inserting with default state",
+                var
+            );
+        }
+        self.vars.get_mut(var)
+    }
+
+    fn get_info(&self, var: &str) -> Option<&VarInfo> {
+        self.vars.get(var)
+    }
+
+    // Retrieve a variable's state
+    fn _get_state(&mut self, var: &str) -> Option<&VarState> {
+        if let Some(info) = self.get_info(var) {
+            Some(&info.state)
+        } else {
+            None
+        }
+    }
+
+    // Retrieve a variable's state
+    fn update_state(&mut self, var: &str, new_state: &VarState) {
+        let info = self.get_info_mut(var);
+        if let Some(info) = info {
+            //info.state = new_state.clone();
+            info.state = *new_state;
+        }
+    }
+
+    fn get_loop_depth(&mut self, name: &str) -> usize {
+        let state = self.get_info(name);
+        if let Some(state) = state {
+            state.depth
+        } else {
+            0
+        }
+    }
+
+    pub fn _tables_are_consistent(&self, other: &StateTbl) -> Result<(), LinearityError> {
+        for (key, value) in &self.vars {
+            if let Some(other_value) = other.vars.get(key) {
+                if value != other_value {
+                    return Err(LinearityError::StateInconsistency {
+                        message: format!(
+                            "Variable '{}' state mismatch: {:?} != {:?}",
+                            key, value, other_value
+                        ),
+                    });
+                }
+            } else {
+                return Err(LinearityError::StateInconsistency {
+                    message: format!("Variable '{}' not found in the other table", key),
+                });
+            }
+        }
+        // Also check for variables in `other` not present in `self`
+        for key in other.vars.keys() {
+            if !self.vars.contains_key(key) {
+                return Err(LinearityError::StateInconsistency {
+                    message: format!("Variable '{}' not found in the first table", key),
+                });
+            }
+        }
+
+        Ok(())
+    }
+}
+
+struct LinearityChecker {
+    //state_tbl: StateTbl,
```

Comment:
why is this commented? can you add a `TODO` with the reason?

Reply:
Explained in impl LinearityChecker review, just bellow. Removed comment